src/modulation.jl

```diff
@@ -0,0 +1,194 @@
+abstract type Modulation end
+abstract type BOC <: Modulation end
+
+struct BOCsin{M <: Union{AbstractFloat, Integer}, N <: Union{AbstractFloat, Integer}} <: BOC
+    m::M
+    n::N
+    BOCsin(m, n) =
+        m >= 1 && n >= 1 ?
+        new{typeof(m), typeof(n)}(m, n) :
+        error("m and n must be >= 1")
+end
+
+struct BOCcos{M <: Union{AbstractFloat, Integer}, N <: Union{AbstractFloat, Integer}} <: BOC
+    m::M
+    n::N
+    BOCcos(m, n) =
+        m >= 1 && n >= 1 ?
+        new{typeof(m), typeof(n)}(m, n) :
+        error("m and n must be >= 1")
+end
+
+struct LOC <: Modulation end
+
+struct CBOC{B1 <: BOC, B2 <: BOC} <: BOC
+    boc1::B1
+    boc2::B2
+    boc1_power::Float32
+    CBOC(boc1, boc2, boc1_power) =
+        0 < boc1_power < 1 && boc1.n == boc2.n ?
+        new{typeof(boc1), typeof(boc2)}(boc1, boc2, boc1_power) :
+        error("Power of BOC1 must be between 0 and 1 and n of both BOCs must match")
+end
+
+get_code_type(system::T) where T <: AbstractGNSS = get_code_type(system, get_modulation(T))
+
+get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation) where T = T
+get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation::CBOC) where T = typeof(modulation.boc1_power)
+
+get_code_factor(system::T) where T <: AbstractGNSS = get_code_factor(get_modulation(T))
+get_code_factor(modulation::LOC) = 1
+get_code_factor(modulation::BOC) = modulation.n
+get_code_factor(modulation::CBOC) = modulation.boc1.n
+
+"""
+$(SIGNATURES)
+Get the spectral power
+"""
+get_code_spectrum(system::T, f) where T <: AbstractGNSS = get_code_spectrum(get_modulation(T), system, f)
+get_code_spectrum(modulation::LOC, system, f) = get_code_spectrum_BPSK(get_code_frequency(system), f)
+get_code_spectrum(modulation::BOCsin, system, f) =
+    get_code_spectrum_BOCsin(modulation.n * get_code_frequency(system), modulation.m * get_code_frequency(system), f)
+get_code_spectrum(modulation::BOCcos, system, f) =
+    get_code_spectrum_BOCcos(modulation.n * get_code_frequency(system), modulation.m * get_code_frequency(system), f)
+get_code_spectrum(modulation::CBOC, system, f) = error("To be implemented")
```

Comment:
I think this should be
```suggestion
get_code_spectrum(modulation::CBOC, system, f) = get_code_spectrum(m.boc1, system, f) * m.boc1_power + get_code_spectrum(m.boc2, system, f) * (1-m.boc1_power)
```